### PR TITLE
Remove SHA Computation from Flatten Blob

### PR DESCRIPTION
**Pull request by anthony-murphy**
This is related to #4310 as to validate data store attributes and fail fast we need synchronous access to the attributes blob. The only thing blocking that is the sha calculation in flattenCore.

It seems safe to replace the SHA with a uuid, as the SHA is only used as an in memory id when creating the snapshotTree. Once buildSnapshotTree is synchronous, it allows the unwinding of a bunch of async code that existed simply to all the async SHA calculation.

### CODE REVIEW

File: packages/loader/driver-utils/src/buildSnapshotTree.ts [hunk 1]
```diff
@@ -14,26 +14,27 @@ import {
...
     const entries: git.ITreeEntry[] = [];
     for (const treeEntry of treeEntries) {
         const subPath = `${path}${treeEntry.path}`;
 
         if (treeEntry.type === TreeEntry.Blob) {
             const blob = treeEntry.value as IBlob;
             const buffer = IsoBuffer.from(blob.contents, blob.encoding);
-            const sha = await gitHashFile(buffer);
-            blobMap.set(sha, buffer.toString("base64"));
+            const id = uuid();
```
Comment (anthony-murphy) [inline packages/loader/driver-utils/src/buildSnapshotTree.ts:31]: This is the crux of the change. As far as i can tell this is safe, as this SHA is really only used as an Id for our internal representation of the snapshot and is never pushed back to the server

File: packages/runtime/container-runtime/src/dataStores.ts [hunk 2]
```diff
@@ -160,22 +160,19 @@ export class DataStores implements IDisposable {
...
 
         const flatBlobs = new Map<string, string>();
-        let flatBlobsP = Promise.resolve(flatBlobs);
-        let snapshotTreeP: Promise<ISnapshotTree> | null = null;
+        let snapshotTree: ISnapshotTree | null = null;
         if (attachMessage.snapshot) {
-            snapshotTreeP = buildSnapshotTree(attachMessage.snapshot.entries, flatBlobs);
-            // flatBlobs' validity is contingent on snapshotTreeP's resolution
-            flatBlobsP = snapshotTreeP.then(() => { return flatBlobs; });
+            snapshotTree = buildSnapshotTree(attachMessage.snapshot.entries, flatBlobs);
```
Comment (anthony-murphy) [inline packages/runtime/container-runtime/src/dataStores.ts:165]: we have the snapshot synchronously here, so by not calculating SHA, there is no need to go async

File: packages/runtime/container-runtime/src/test/dataStoreContext.spec.ts [hunk 3]
```diff
@@ -225,9 +225,9 @@ describe("Data Store Context Tests", () => {
 
             remotedDataStoreContext = new RemotedFluidDataStoreContext(
                 dataStoreId,
-                Promise.resolve(snapshotTree),
+                snapshotTree,
```
Comment (anthony-murphy) [inline packages/runtime/container-runtime/src/test/dataStoreContext.spec.ts:228]: sooo much promise wrapping can be removed, as these things are always synchronously avaialable, it's just the SHA that made them async

File: packages/loader/driver-utils/src/buildSnapshotTree.ts (hunk 1 above)
Comment (vladsud) [reply packages/loader/driver-utils/src/buildSnapshotTree.ts:31]: I believe you are right! We can go even cheaper - how about static counter? :)

File: packages/loader/driver-utils/src/buildSnapshotTree.ts [hunk 4]
```diff
@@ -14,26 +14,27 @@ import {
...
+            blobMap.set(id, buffer.toString("base64"));
 
             const entry: git.ITreeEntry = {
                 mode: FileMode[treeEntry.mode],
                 path: subPath,
-                sha,
+                sha: id,
```
Comment (vladsud) [inline packages/loader/driver-utils/src/buildSnapshotTree.ts:37]: How hard is it to remove git lingvo from here? Using git interfaces here is very confusing and is not needed, as far as I can tell

**Review by vladsud — APPROVED** [review]
Thanks, great change - I wanted to do the same! :)

**Comment by vladsud** [comment]
Related: https://github.com/microsoft/FluidFramework/issues/224
I thought I opened issue for this kind of work, but I was not able to find it

File: packages/loader/driver-utils/src/buildSnapshotTree.ts (hunk 4 above)
Comment (anthony-murphy) [reply packages/loader/driver-utils/src/buildSnapshotTree.ts:37]: I agree. i briefly looked at it, but quickly ended up in protocol definitions which is harder to stage. I'll file an issue as i agree we shouldn't need git on the on client side beyond maybe a driver that depends on it.

**Pull request merged**